### ISSUE TITLE
_tkinter.TclError: integer value too large to represent

### ISSUE THREAD

**Issue by Justinjay282** [Justinjay282/qwerty#44]
Hi sir @WK-GiHu, I currently trying to create the RFID with register.
But I'm encountering this error. 
```
Exception in thread Thread-2:
Traceback (most recent call last):
  File "/usr/lib/python3.7/threading.py", line 917, in _bootstrap_inner
    self.run()
  File "/home/pi/Kiosk/RFIDDevice.py", line 51, in run
    self.app.event_generate('<<RFID>>', when='tail', state=str(RFID_NUMBER))
  File "/usr/lib/python3.7/tkinter/__init__.py", line 1673, in event_generate
    self.tk.call(args)
_tkinter.TclError: integer value too large to represent
```
I'm doing it like the structure of Fingerprint but much simpler because I will not get if the RFID is already registered or not even if it's registered the program will just update it if the RFID to be scanned is not yet registered. Unfortunately, I got that error... that's why I can't proceed...

I added new file in the repository: `Kiosk_Main_RFID.py`, `RFIDDevice` and `Update_Residents_RFID.py`.

**Comment by WK-GiHu** [comment]
@Justinjay282 

>`state=str(RFID_NUMBER))`
`_tkinter.TclError: integer value too large to represent`

Yes, unfortunately  `state=` is limeted.

---

>`AttributeError: 'Event' object has no attribute 'data'`

To use `data=` you have to register `%d`

**.bind** changes!
```
        self.app.tk.call("bind", self.app, "<<RFID>>", self.app.register(callback) + " %d")
```
**callback** changes!
```
    def on_rfid(self, data):
```
 No more `event` parameter
Therefore, this will not work
```
            if event.result: 
                event.state = 1
                return self.on_grant_access(event)
```

change to
```
    def on_rfid(self, data):
        class Event: pass

        self.cursor.execute("SELECT * FROM residents_admin WHERE RFID = %s", data)
        result = self.cursor.fetchone()
        if result: 
            event = Event()
            event.state = 11
            return self.on_grant_access(event)
```

**Comment by Justinjay282** [comment]
sir @WK-GiHu,
> @Justinjay282
> 
> > `state=str(RFID_NUMBER))`
> > `_tkinter.TclError: integer value too large to represent`
> 
> Yes, unfortunately `state=` is limeted.
> 
> > `AttributeError: 'Event' object has no attribute 'data'`
> 
> To use `data=` you have to register `%d`
> 
Yes sir,
I've read in stack overflow that it'll not work since an interpreter is needed to connect the python and tcl. It's actually confusing me more...

> **.bind** changes!
> 
> ```
>         self.app.tk.call("bind", self.app, "<<RFID>>", self.app.register(callback) + " %d")
> ```
> 

I will changed the `   
``` 
self.fp.bind('SEARCH', self.on_fingerprint)`
```
to this? 
```
self.app.tk.call("bind", self.app, "<<RFID>>", self.app.register(callback) + " %d")
```

**Comment by WK-GiHu** [comment]
@Justinjay282 

>I will changed the `
>`self.fp.bind('SEARCH', self.on_fingerprint)`

Don't touch `Fingerprint`

---

>but I'm having an error

You didn't show the error, but i assume `self._bind_funcid` becomes `None`.
Else show the error!
Split it into two lines:
```
        self._bind_funcid =  self.register(self.app.register(callback))
        self.bind("<<RFID>>", self._bind_funcid + " %d")
```

**Comment by Justinjay282** [comment]
sir @WK-GiHu, I've edited the `RFIDDevice.py` I adapt this.
```
        self._bind_funcid =  self.register(self.app.register(callback))
        self.bind("<<RFID>>", self._bind_funcid + " %d")
```

and this is the error occurred:
```
Exception in thread Thread-2:
Traceback (most recent call last):
  File "/usr/lib/python3.7/threading.py", line 917, in _bootstrap_inner
    self.run()
  File "/home/pi/Kiosk/RFIDDevice.py", line 52, in run
    self.app.event_generate('<<RFID>>', when='tail', data=RFID_NUMBER)
  File "/usr/lib/python3.7/tkinter/__init__.py", line 1673, in event_generate
    self.tk.call(args)
RuntimeError: main thread is not in main loop
```

**Comment by WK-GiHu** [comment]
@Justinjay282 

>`RuntimeError: main thread is not in main loop`

That's unrelated to `self.bind(...`
You have broken how to start the background task.
I assume `self.app` is not in the `.mainloop()`

**Comment by Justinjay282** [comment]
sir @WK-GiHu, I can't locate what I did wrong... I have `self.app = app` in `__init__`
in background task. before `while True` I declare the RFID reader first before the scan forever.

**Comment by WK-GiHu** [comment]
@Justinjay282 

>I can't locate what I did wrong..

OK, let's verify.
Copy form `FingerprintDevice.py` the `if __name__ ...` block into `RFIDDevice.py`.
Adapt accordingly for **RFID**, run `RFIDDevice.py` and verify if the error persists.

**Comment by Justinjay282** [comment]
sir @WK-GiHu, I've edited the `RFIDDevice.py` and put the `if __name_ ...` from fingerprint. 
I'm still receiving this error. 
```
Exception in Tkinter callback
Traceback (most recent call last):
  File "/usr/lib/python3.7/tkinter/__init__.py", line 1705, in __call__
    return self.func(*args)
  File "/home/pi/Kiosk/RFIDDevice.py", line 66, in on_rfid
    print('on_RFID()  positionNumber={}'.format(event.data))
AttributeError: 'str' object has no attribute 'data'
```
different from when I run it in may `Kiosk_Main`

**Comment by WK-GiHu** [comment]
@Justinjay282 

Still one Liner, thus unbind will fail, should bei https://github.com/Justinjay282/qwerty/issues/44#issuecomment-606689036
https://github.com/Justinjay282/qwerty/blob/f6c3652de91062d7164a56bb35c5864b3e461e3e/RFIDDevice.py#L27

---

Thus ist evil, group all import tkinter
https://github.com/Justinjay282/qwerty/blob/f6c3652de91062d7164a56bb35c5864b3e461e3e/Update_Residents_RFID.py#L1-L2
https://github.com/Justinjay282/qwerty/blob/f6c3652de91062d7164a56bb35c5864b3e461e3e/Kiosk_Main_RFID.py#L4-L5

**Comment by Justinjay282** [comment]
sir @WK-GiHu, I'm receiving an error.
```
/usr/local/lib/python3.7/dist-packages/mfrc522/MFRC522.py:151: RuntimeWarning: This channel is already in use, continuing anyway.  Use GPIO.setwarnings(False) to disable warnings.
  GPIO.setup(pin_rst, GPIO.OUT)
Traceback (most recent call last):
  File "/home/pi/Kiosk/RFIDDevice.py", line 86, in <module>
    rf.bind('SEARCH', on_rfid)
  File "/home/pi/Kiosk/RFIDDevice.py", line 27, in bind
    self._bind_funcid =  self.register(self.app.register(callback))
AttributeError: 'RFIDThread' object has no attribute 'register'
```
When I changed this line https://github.com/Justinjay282/qwerty/blob/f6c3652de91062d7164a56bb35c5864b3e461e3e/RFIDDevice.py#L27


Should I remove the `*`? and then just import is as `import tkinter as tk`?

**Comment by WK-GiHu** [comment]
@Justinjay282 

>’AttributeError: 'RFIDThread' object has no attribute 'register'’

That's correct, my bad, copy from edited https://github.com/Justinjay282/qwerty/issues/44#issuecomment-606689036 as is.

**Comment by Justinjay282** [comment]
sir @WK-GiHu, I tried it but It's not working... I don't know why. That's why I'm having a trouble. Tried so many things but Can''t really solve it :(

**Comment by WK-GiHu** [comment]
@Justinjay282 

please update RFIDDevice.py, still wrong...
https://github.com/Justinjay282/qwerty/blob/899f9b08ddf39c99566088ff1408d4f7ac4dd4f0/RFIDDevice.py#L27-L28

---

>I wondering why after I scanned an RFID. I can't scan anymore RFID

This will fail, as you change to `str(...`
https://github.com/Justinjay282/qwerty/blob/899f9b08ddf39c99566088ff1408d4f7ac4dd4f0/RFIDDevice.py#L48

Remove the whole, it's useless
https://github.com/Justinjay282/qwerty/blob/899f9b08ddf39c99566088ff1408d4f7ac4dd4f0/RFIDDevice.py#L43-L46

Change to `if self.rfid ...`

---


Verify, that you get the following print...
https://github.com/Justinjay282/qwerty/blob/899f9b08ddf39c99566088ff1408d4f7ac4dd4f0/Update_Residents_RFID.py#L142-L144

**Comment by Justinjay282** [comment]
sir @WK-GiHu, I've edited the `RFIDDevice.py` 

>Verify, that you get the following print...

Yes sir, the `print('bind REGISTER') is printing.
when I closed the window. the `print('bind SEARCH') is also printing.

**Comment by WK-GiHu** [comment]
@Justinjay282 

Still double work and will fail if the first read is None.
Remove all `RFID_NUMBER` and replace with `self.id`.
Remove Line 44 to 45
https://github.com/Justinjay282/qwerty/blob/0fb483ba741c1841e238fdc9d6671cb2895f2036/RFIDDevice.py#L44-L48

---

Add `print('on_rfid({})'.format(data))` after line 104 to verify the event get fired.
You should see:
1. bind REGISTER
2. Ready RFID
3. on_rfid(?????????)
https://github.com/Justinjay282/qwerty/blob/0fb483ba741c1841e238fdc9d6671cb2895f2036/Update_Residents_RFID.py#L104

---

Looks wrong, why table '...admin'
https://github.com/Justinjay282/qwerty/blob/0fb483ba741c1841e238fdc9d6671cb2895f2036/Update_Residents_RFID.py#L110

**Comment by Justinjay282** [comment]
sir @WK-GiHu, I've edited the `RFIDDevice.py` and `Update_Residents_RFID.py`.
I'm sorry but what I told you is incorrect. 
I think The error is still in `RFIDDevice.py` since when I run it standalone when I clicked the button register It's only scanning once and need to click the button again to register...

**Comment by WK-GiHu** [comment]
@Justinjay282 

>need to click the button again to register...

That's how you implemented the behavior in REGISTER mode, to read only once and wait until the next `.bind('REGISTER', ...`.

Do you need a second one to compare with the first read, like in FingerprintDevice?

**Comment by Justinjay282** [comment]
sir @WK-GiHu, No sir. I just want to scan continuously if the RFID Card is already registered then I can scan again no need to back then click the register RFID again. Just like how the login works, The user can scan RFID continuously even if the RFID is not yet registered.

**Comment by WK-GiHu** [comment]
@Justinjay282 

You close the Toplevel only after UPDATE.
Why do you use UPDATE instead of INSERT?

https://github.com/Justinjay282/qwerty/blob/4aafa475ca3473d796c4983feb396288d5c1d999/Update_Residents_RFID.py#L114-L117

you should destroy after every Event callback, so outident line 117

**Comment by Justinjay282** [comment]
> @Justinjay282
> 
> You close the Toplevel only after UPDATE.
> Why do you use UPDATE instead of INSERT?
> 

sir @WK-GiHu, I used UPDATE since the admin will just select a details are already existing and I just need to put `RFID number` to `RFID column` in the database. 
> https://github.com/Justinjay282/qwerty/blob/4aafa475ca3473d796c4983feb396288d5c1d999/Update_Residents_RFID.py#L114-L117
> 
> you should destroy after every Event callback, so outident line 117

In this I've indented the `self.toplevel_rfid.destroy()` so it will just destroy if the "RFID is successfully registered" and not if the "RFID is already registered" so if the RFID is not yet registered the window will not close and the Register will continue registering until the program accepts an RFID I just have to continuously scan another RFID without clicking the `Register RFID` button again. 

but I think what you recommended is enough. if the RFID is  already registered I just have to automatically close the window then let the `admin` try again.

**Comment by WK-GiHu** [comment]
@Justinjay282 

>if the RFID is not yet registered the window will not close and the Register will continue registering 

Add after line 112, `.bind(...`
https://github.com/Justinjay282/qwerty/blob/4aafa475ca3473d796c4983feb396288d5c1d999/Update_Residents_RFID.py#L112

https://github.com/Justinjay282/qwerty/blob/4aafa475ca3473d796c4983feb396288d5c1d999/Update_Residents_RFID.py#L143

OR

disable the following code block to disable to wait at all:
https://github.com/Justinjay282/qwerty/blob/4aafa475ca3473d796c4983feb396288d5c1d999/RFIDDevice.py#L46-L49

**Comment by Justinjay282** [comment]
sir @WK-GiHu, Thank you so much sir. I chose the first one. and put `self.app.rf.bind('REGISTER', self.on_rfid) ` in line 112 https://github.com/Justinjay282/qwerty/blob/4aafa475ca3473d796c4983feb396288d5c1d999/Update_Residents_RFID.py#L112

and It's working fine now sir :) Thank you so much sir!

I just want to share you something sir. Because of COVID19 there's a chance that our thesis is cancel... I don't know what to do actually since this is the last thing needs to be done. We're good as finish in our thesis. We'll just want to wait for the final announcement... since we don't wanna waste what we spent and also the effort.

Sir, also, can I also ask some pointers? I currently don't know where to start studying... since I'm finish now. after converting this to controller I'll start studying programming right away. our admin system made in python is also good now and I want to prepare notes I need to study so I can't waster anymore time.
by the way sir thank you so much! :) .
